watertap/unit_models/reverse_osmosis_base.py

```diff
@@ -0,0 +1,822 @@
+            bounds=lambda b, t, x, p, j: (1e-4, 3e-2)
+            if j in solvent_set
+            else (1e-8, 1e-3),
+            units=units_meta("mass")
+            * units_meta("length") ** -2
+            * units_meta("time") ** -1,
+            doc="Mass flux across membrane at inlet and outlet",
+        )
+
+        @self.Constraint(
+            self.flowsheet().config.time,
+            self.difference_elements,
+            self.config.property_package.phase_list,
+            self.config.property_package.component_list,
+            doc="Solvent and solute mass flux",
+        )
+        def eq_flux_mass(b, t, x, p, j):
+            prop_feed = b.feed_side.properties[t, x]
+            prop_perm = b.permeate_side[t, x]
+            interface = b.feed_side.properties_interface[t, x]
+            comp = self.config.property_package.get_component(j)
+            if comp.is_solvent():
+                return b.flux_mass_phase_comp[t, x, p, j] == b.A_comp[
+                    t, j
+                ] * b.dens_solvent * (
+                    (prop_feed.pressure - prop_perm.pressure)
+                    - (
+                        interface.pressure_osm_phase[p]
+                        - prop_perm.pressure_osm_phase[p]
+                    )
+                )
+            elif comp.is_solute():
+                return b.flux_mass_phase_comp[t, x, p, j] == b.B_comp[t, j] * (
+                    interface.conc_mass_phase_comp[p, j]
+                    - prop_perm.conc_mass_phase_comp[p, j]
+                )
+
+        @self.Expression(
+            self.flowsheet().config.time,
+            self.config.property_package.phase_list,
+            self.config.property_package.component_list,
+            doc="Average flux expression",
+        )
+        def flux_mass_phase_comp_avg(b, t, p, j):
+            return (
+                sum(
+                    b.flux_mass_phase_comp[t, x, p, j] for x in self.difference_elements
+                )
+                / self.nfe
+            )
+
+        if (
+            self.config.concentration_polarization_type
+            == ConcentrationPolarizationType.calculated
+        ):
+
+            @self.Constraint(
+                self.flowsheet().config.time,
+                self.difference_elements,
+                solute_set,
+                doc="Concentration polarization",
+            )
+            def eq_concentration_polarization(b, t, x, j):
+                jw = b.flux_mass_phase_comp[t, x, "Liq", "H2O"] / self.dens_solvent
+                js = b.flux_mass_phase_comp[t, x, "Liq", j]
+                return b.feed_side.properties_interface[t, x].conc_mass_phase_comp[
+                    "Liq", j
+                ] == b.feed_side.properties[t, x].conc_mass_phase_comp["Liq", j] * exp(
+                    jw / self.feed_side.K[t, x, j]
+                ) - js / jw * (
+                    exp(jw / self.feed_side.K[t, x, j]) - 1
+                )
```

Comment:
Ok so this concentration polarization constraint is added on the unit model and the constraint for the other concentration polarization types (i.e. fixed, none) are on the membrane channel? I don't know if there is a better solution.

What's the downside with moving the constraints for the other concentration polarization types to the unit model? I understand that the channel initialization won't have concentration polarization constraint when it could have for fixed and none options, but for the calculated option the channel initialization we would have to assume the interface is at the same state as the bulk - which is also fine for the fixed and none options.

Reply:
@adam-a-a discussed this and thought this was the best, though not great, solution.

Putting the flux on the channel doesn't make much sense as it is the mass transfer term. This constraint is also different on different sides of the channel. However, all the other concentration polarization equations are the same (as far as I understand) on both sides of the channel, so conceptually this separation makes some sense. In any case, we now create a `cp_modulus` Var in every case except `none`, so in some ways the channel doesn't really care beyond "create cp_modulus or not" -- and this constraint serves to calculate it.

We could put them all the unit model; however, note that any equations we put on the unit model need to be re-implemented in each unit.

Reply:
Does this suggest we should have a common utility function/method for creating these constraints at the unit level then (e.g. on a common base class for the models)?

Reply:
@TimBartholomew I think we now only have this `eq_concentration_polarization` at the unit level when CP type is calculated. When none or fixed, we now only have a cp_modulus constraint in the membrane channel CV which uses the cp_modulus to get interface concentration. When CP type is none, concentrations at bulk and interface are set equal using `eq_cp_modulus`.

Reply:
I think what we have now is fine actually. In the membrane channel CV, we just use a cp_modulus constraint, which makes sense from a channel perspective (i.e., each channel will have its own cp_modulus w/respect to length across the channel, and this is just the ratio of interface to bulk concentration). 

However, the constraint for calculating the membrane interface concentration can take on different forms, depending on whether we are using a first principles relationship, an empirical relationship, or depending on _which_ channel and flow configuration we are considering. Thus, it makes sense for any detailed calculation for CP to be at the unit model; meanwhile, we do no such calculation at the channel and merely compute interface concentration be an assumed ratio. 

Overall, we no longer have the same constraint living in either the unit model or channel model when switching between Enum types for concentration polarization. `eq_concentration_polarization` only lives on the unit model and `eq_cp_modulus` only lives on the channel CV.